lib/rspec/support/method_signature_verifier.rb

```diff
@@ -85,12 +85,58 @@ def has_kw_args_in?(args)
             (args.last.empty? || args.last.keys.any? { |x| x.is_a?(Symbol) })
         end
 
-        # Without considering what the last arg is, could it
```

Comment:
I have a faint feeling that the incorrect behaviour is in the `has_kw_args_in?` above, not related to the `could_contain_kw_args?` call.

With the test case that reproduces the problem from #522,
```ruby
            it 'treats the final positional argument as a hash' do
              expect(valid?(1, { foo: 'bar', baz: 'eggs'})).to eq(true)
            end
```

The problem is a call to `arity_kw` defined as:
```ruby
def arity_kw(x, y = {}, z:2); end
```
 - **may** contain kwargs (so arguably `could_contain_kw_args?` returning `true` is correct)
 - has a variable number of positional arguments

The last argument can be matched to either `y` or considered a kwargs depending on what this last argument is.

With `{ foo: 'bar', baz: 'eggs'}` specifically, we can really detect it's a kwarg by checking if:
1. it is a Hash
2. all of its keys are Symbols
3. **all** of its keys of the arg are in the method signature as kwargs

With `foo` and `bar` as the last arg keys and `z` as the only kwarg in the method signature, we can certainly tell that the last argument is **not** a kwarg.

Reply:
Sorry for the lengthy comment, I've tried to refresh my memory by writing things down. Glancing over, I see that your implementation is addressing this exact problem in `has_extra_keys`.

Reply:
Yep this PR is entirely about addressing ambiguity where traditional style hashes can be keywords when there is this duality.